commit 34c8d3ffb64883b9279c03e79c1b08c9a1931940
Author: Nicholas Piggin <npiggin@gmail.com>
Date:   Fri Nov 10 04:27:37 2017 +1100

    powerpc/64s/hash: Fix fork() with 512TB process address space
    
    commit effc1b25088502fbd30305c79773de2d1f7470a6 upstream.
    
    Hash unconditionally resets the addr_limit to default (128TB) when the
    mm context is initialised. If a process has > 128TB mappings when it
    forks, the child will not get the 512TB addr_limit, so accesses to
    valid > 128TB mappings will fail in the child.
    
    Fix this by only resetting the addr_limit to default if it was 0. Non
    zero indicates it was duplicated from the parent (0 means exec()).
    
    Fixes: f4ea6dcb08ea ("powerpc/mm: Enable mappings above 128TB")
    Signed-off-by: Nicholas Piggin <npiggin@gmail.com>
    Reviewed-by: Aneesh Kumar K.V <aneesh.kumar@linux.vnet.ibm.com>
    Signed-off-by: Michael Ellerman <mpe@ellerman.id.au>
    Signed-off-by: Greg Kroah-Hartman <gregkh@linuxfoundation.org>

diff --git a/arch/powerpc/mm/mmu_context_book3s64.c b/arch/powerpc/mm/mmu_context_book3s64.c
index 05e15386d4cb..b94fb62e60fd 100644
--- a/arch/powerpc/mm/mmu_context_book3s64.c
+++ b/arch/powerpc/mm/mmu_context_book3s64.c
@@ -93,11 +93,11 @@ static int hash__init_new_context(struct mm_struct *mm)
 		return index;
 
 	/*
-	 * We do switch_slb() early in fork, even before we setup the
-	 * mm->context.addr_limit. Default to max task size so that we copy the
-	 * default values to paca which will help us to handle slb miss early.
+	 * In the case of exec, use the default limit,
+	 * otherwise inherit it from the mm we are duplicating.
 	 */
-	mm->context.addr_limit = DEFAULT_MAP_WINDOW_USER64;
+	if (!mm->context.addr_limit)
+		mm->context.addr_limit = DEFAULT_MAP_WINDOW_USER64;
 
 	/*
 	 * The old code would re-promote on fork, we don't do that when using
diff -uN a/1.txt b/1.txt
--- a/dummy/rpi_1169_34c8d3ffb64883b9279c03e79c1b08c9a1931940.txt	1970-01-01 00:00:00.000000000 +0000
+++ b/dummy/rpi_1169_34c8d3ffb64883b9279c03e79c1b08c9a1931940.txt	2013-12-23 04:07:40.000000000 +0000
@@ -0,0 +1 @@
+dummy file to ensure patch has content.
